tests/src/Interop/PInvoke/CriticalHandles/Test/Test.cs

```diff
@@ -0,0 +1,393 @@
+    }
+
+    public static void Out()
+    {
+        IntPtr handleValue = new IntPtr(3);
+        AbstractCriticalHandle handle;
+        Assert.Throws<MarshalDirectiveException>(() => Native.Out(handleValue, out handle), "Calling P/Invoke that has an out abstract critical handle parameter");
+    }
+
+    public static void InRef()
+    {
+        IntPtr handleValue = new IntPtr(4);
+        AbstractCriticalHandle handle = new CriticalHandleWithNoDefaultCtor(handleValue);
+        Native.InRef(ref handle, null);
+        Assert.AreEqual(handleValue.ToInt32(), handle.Handle.ToInt32(), "Handle value");
+    }
+
+    public static void Ref()
+    {
+        IntPtr handleValue = new IntPtr(5);
+        AbstractCriticalHandle handle = new CriticalHandleWithNoDefaultCtor(handleValue);
+        Assert.Throws<MarshalDirectiveException>(() => Native.Ref(ref handle, null), "Calling P/Invoke that has a ref abstract critical handle parameter");
+    }
+
+    internal class Native
+    {
+        [UnmanagedFunctionPointer(System.Runtime.InteropServices.CallingConvention.StdCall)]
+        [return: MarshalAs(UnmanagedType.Bool)]
+        internal delegate bool HandleCallback(IntPtr handle);
+
+        [DllImport("CriticalHandlesNative.dll", CallingConvention = CallingConvention.StdCall)]
+        internal static extern IntPtr In(AbstractCriticalHandle handle, HandleCallback handleCallback);
+
+        [DllImport("CriticalHandlesNative.dll", CallingConvention = CallingConvention.StdCall)]
+        internal static extern void Out(IntPtr handleValue, out AbstractCriticalHandle handle);
+
+        [DllImport("CriticalHandlesNative.dll", EntryPoint = "Ref", CallingConvention = CallingConvention.StdCall)]
+        internal static extern IntPtr InRef([In]ref AbstractCriticalHandle handle, HandleCallback handleCallback);
+
+        [DllImport("CriticalHandlesNative.dll", CallingConvention = CallingConvention.StdCall)]
+        internal static extern IntPtr Ref(ref AbstractCriticalHandle handle, HandleCallback handleCallback);
+
+        [DllImport("CriticalHandlesNative.dll", CallingConvention = CallingConvention.StdCall)]
+        internal static extern AbstractCriticalHandle Ret(IntPtr handleValue);
+    }
+}
+
+public class NoDefaultCtorCriticalHandleTest
+{
+    public static void In()
+    {
+        IntPtr handleValue = new IntPtr(1);
+        CriticalHandleWithNoDefaultCtor handle = new CriticalHandleWithNoDefaultCtor(handleValue);
+        IntPtr value;
+        value = Native.In(handle, null);
+        Assert.AreEqual(handleValue.ToInt32(), value.ToInt32(), "Handle value");
+    }
+
+    public static void Ret()
+    {
+        IntPtr handleValue = new IntPtr(2);
+        //TODO: Expected MissingMemberException but throws MissingMethodException
+        Assert.Throws<MissingMethodException>(() => Native.Ret(handleValue), "Calling P/Invoke that returns an no default ctor critical handle");
```

Comment:
@AaronRobinsonMSFT Need confirm here which exception should throw

Reply:
I believe @hughbe was also reporting this in a test in corefx. Can you confirm that @hughbe?

Can you check if we have this coverage in corefx? If so, please remove here.

Reply:
@AaronRobinsonMSFT I did not found similar test in CoreFx., @hughbe do you have any idea?

Reply:
I didn’t. All I can think of is something to do with handles and getting a variant from them not being allowed